### ISSUE TITLE
inconsistent dna"" - DNASequence() behavior

### ISSUE THREAD

**Issue by cjprybol** [BioJulia/Bio.jl#348]
This is an example of simulating RNA transcripts (in DNA format) from a fake transcriptome and then randomly making sequencing errors. There are two for loops, the first utilizes `dna""` to create a new, empty dna sequence and in the second for loop I use the `DNASequence()` function.

tracing through the steps with `@show`,  `dna""` doesn't seem to release the object in memory from the prior loop. This causes the `new_seq` variable to be continuously extended during the for loop iterations. `DNASequence()` has the behavior I expected, and assigns an empty sequence to the variable `new_seq` during each for loop.

The key lines are
```
@show new_seq = dna""
```
vs.
```
@show new_seq = DNASequence()
```

Is this intended behavior, or potentially a bug in the parsing of the `dna""` constructor?

**code**
```
using Bio.Seq
using StatsBase

srand(1);
transcriptome = [randdnaseq(rand(8:12)) for i=1:3];
number_of_reads = 5;
answers = sample(transcriptome, number_of_reads);

error = .10;
weights = WeightVec([1-error, error * 1/3, error * 1/3, error * 1/3]);

sequenced_reads_dna = []
for seq in answers
	@show new_seq = dna""
	for nuc in seq
		@show choice = sample(1:4, weights)
		if choice == 1
			@show push!(new_seq, nuc)
		elseif choice == 2
			# deletion, do nothing
			@show new_seq
		elseif choice == 3
			# mutation
			@show new_seq *= randdnaseq(1)
		elseif choice == 4
			# insertion
			@show new_seq *= push!(randdnaseq(1), nuc)
		end
	end
	push!(sequenced_reads_dna, new_seq)
end

sequenced_reads_function = []
for seq in answers
	@show new_seq = DNASequence()
	for nuc in seq
		@show choice = sample(1:4, weights)
		if choice == 1
			@show push!(new_seq, nuc)
		elseif choice == 2
			# deletion, do nothing
			@show new_seq
		elseif choice == 3
			# mutation
			@show new_seq *= randdnaseq(1)
		elseif choice == 4
			# insertion
			@show new_seq *= push!(randdnaseq(1), nuc)
		end
	end
	push!(sequenced_reads_function, new_seq)
end

answers
sequenced_reads_dna
sequenced_reads_function
```

**REPL**
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.1-pre+2 (2016-09-20 03:34 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit f0d40ec* (22 days old release-0.5)
|__/                   |  x86_64-apple-darwin16.1.0

julia> using Bio.Seq

julia> using StatsBase

julia> srand(1);

julia> transcriptome = [randdnaseq(rand(8:12)) for i=1:3];

julia> number_of_reads = 5;

julia> answers = sample(transcriptome, number_of_reads);

julia> error = .10;

julia> weights = WeightVec([1-error, error * 1/3, error * 1/3, error * 1/3]);

julia> sequenced_reads_dna = []
0-element Array{Any,1}

julia> for seq in answers
               @show new_seq = dna""
               for nuc in seq
                       @show choice = sample(1:4, weights)
                       if choice == 1
                               @show push!(new_seq, nuc)
                       elseif choice == 2
                               # deletion, do nothing
                               @show new_seq
                       elseif choice == 3
                               # mutation
                               @show new_seq *= randdnaseq(1)
                       elseif choice == 4
                               # insertion
                               @show new_seq *= push!(randdnaseq(1), nuc)
                       end
               end
               push!(sequenced_reads_dna, new_seq)
       end
new_seq = @dna_str("") = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
A
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
AT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 3nt DNA Sequence:
ATA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 4nt DNA Sequence:
ATAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 5nt DNA Sequence:
ATAGC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
ATAGCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
ATAGCAA
choice = sample(1:4,weights) = 3
new_seq *= randdnaseq(1) = 8nt DNA Sequence:
ATAGCAAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 9nt DNA Sequence:
ATAGCAAAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 10nt DNA Sequence:
ATAGCAAAAG
new_seq = @dna_str("") = 7nt DNA Sequence:
ATAGCAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
ATAGCAAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 9nt DNA Sequence:
ATAGCAAAT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 10nt DNA Sequence:
ATAGCAAATA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 11nt DNA Sequence:
ATAGCAAATAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 12nt DNA Sequence:
ATAGCAAATAGC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 13nt DNA Sequence:
ATAGCAAATAGCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 14nt DNA Sequence:
ATAGCAAATAGCAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 15nt DNA Sequence:
ATAGCAAATAGCAAT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 16nt DNA Sequence:
ATAGCAAATAGCAATA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 17nt DNA Sequence:
ATAGCAAATAGCAATAG
new_seq = @dna_str("") = 17nt DNA Sequence:
ATAGCAAATAGCAATAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 18nt DNA Sequence:
ATAGCAAATAGCAATAGT
choice = sample(1:4,weights) = 3
new_seq *= randdnaseq(1) = 19nt DNA Sequence:
ATAGCAAATAGCAATAGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 20nt DNA Sequence:
ATAGCAAATAGCAATAGTCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 21nt DNA Sequence:
ATAGCAAATAGCAATAGTCAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 22nt DNA Sequence:
ATAGCAAATAGCAATAGTCACA
choice = sample(1:4,weights) = 4
new_seq *= push!(randdnaseq(1),nuc) = 24nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 25nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 26nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCTA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 27nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCTAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 28nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCTAGT
choice = sample(1:4,weights) = 4
new_seq *= push!(randdnaseq(1),nuc) = 30nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCTAGTGC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 31nt DNA Sequence:
ATAGCAAATAGCAATAGTCACAGCTAGTGCT
new_seq = @dna_str("") = 18nt DNA Sequence:
ATAGCAAATAGCAATAGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 19nt DNA Sequence:
ATAGCAAATAGCAATAGTT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 20nt DNA Sequence:
ATAGCAAATAGCAATAGTTC
choice = sample(1:4,weights) = 3
new_seq *= randdnaseq(1) = 21nt DNA Sequence:
ATAGCAAATAGCAATAGTTCC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 22nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 23nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 24nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 25nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 26nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACTA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 27nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACTAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 28nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACTAGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 29nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACTAGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 30nt DNA Sequence:
ATAGCAAATAGCAATAGTTCCCACTAGTCT
new_seq = @dna_str("") = 20nt DNA Sequence:
ATAGCAAATAGCAATAGTTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 21nt DNA Sequence:
ATAGCAAATAGCAATAGTTCT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 22nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 23nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 24nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 25nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 26nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 27nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 28nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACTA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 29nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACTAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 30nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACTAGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 31nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACTAGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 32nt DNA Sequence:
ATAGCAAATAGCAATAGTTCTCACACTAGTCT

julia> sequenced_reads_function = []
0-element Array{Any,1}

julia> for seq in answers
               @show new_seq = DNASequence()
               for nuc in seq
                       @show choice = sample(1:4, weights)
                       if choice == 1
                               @show push!(new_seq, nuc)
                       elseif choice == 2
                               # deletion, do nothing
                               @show new_seq
                       elseif choice == 3
                               # mutation
                               @show new_seq *= randdnaseq(1)
                       elseif choice == 4
                               # insertion
                               @show new_seq *= push!(randdnaseq(1), nuc)
                       end
               end
               push!(sequenced_reads_function, new_seq)
       end
new_seq = DNASequence() = 0nt DNA Sequence:

choice = sample(1:4,weights) = 2
new_seq = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
T
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
TA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 3nt DNA Sequence:
TAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 4nt DNA Sequence:
TAGC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 5nt DNA Sequence:
TAGCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
TAGCAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
TAGCAAT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
TAGCAATA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 9nt DNA Sequence:
TAGCAATAG
new_seq = DNASequence() = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
A
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
AT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 3nt DNA Sequence:
ATA
choice = sample(1:4,weights) = 2
new_seq = 3nt DNA Sequence:
ATA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 4nt DNA Sequence:
ATAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 5nt DNA Sequence:
ATACA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
ATACAA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
ATACAAT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
ATACAATA
choice = sample(1:4,weights) = 2
new_seq = 8nt DNA Sequence:
ATACAATA
new_seq = DNASequence() = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
T
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
TC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 3nt DNA Sequence:
TCA
choice = sample(1:4,weights) = 4
new_seq *= push!(randdnaseq(1),nuc) = 5nt DNA Sequence:
TCATC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
TCATCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
TCATCAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
TCATCACT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 9nt DNA Sequence:
TCATCACTA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 10nt DNA Sequence:
TCATCACTAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 11nt DNA Sequence:
TCATCACTAGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 12nt DNA Sequence:
TCATCACTAGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 13nt DNA Sequence:
TCATCACTAGTCT
new_seq = DNASequence() = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
T
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
TC
choice = sample(1:4,weights) = 4
new_seq *= push!(randdnaseq(1),nuc) = 4nt DNA Sequence:
TCGA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 5nt DNA Sequence:
TCGAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
TCGACA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
TCGACAC
choice = sample(1:4,weights) = 2
new_seq = 7nt DNA Sequence:
TCGACAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
TCGACACA
choice = sample(1:4,weights) = 4
new_seq *= push!(randdnaseq(1),nuc) = 10nt DNA Sequence:
TCGACACAGG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 11nt DNA Sequence:
TCGACACAGGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 12nt DNA Sequence:
TCGACACAGGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 13nt DNA Sequence:
TCGACACAGGTCT
new_seq = DNASequence() = 0nt DNA Sequence:

choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 1nt DNA Sequence:
T
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 2nt DNA Sequence:
TC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 3nt DNA Sequence:
TCA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 4nt DNA Sequence:
TCAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 5nt DNA Sequence:
TCACA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 6nt DNA Sequence:
TCACAC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 7nt DNA Sequence:
TCACACT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 8nt DNA Sequence:
TCACACTA
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 9nt DNA Sequence:
TCACACTAG
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 10nt DNA Sequence:
TCACACTAGT
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 11nt DNA Sequence:
TCACACTAGTC
choice = sample(1:4,weights) = 1
push!(new_seq,nuc) = 12nt DNA Sequence:
TCACACTAGTCT

julia> answers
5-element Array{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},1}:
 ATAGCAATAG
 ATAGCAATAG
 TCACACTAGTCT
 TCACACTAGTCT
 TCACACTAGTCT

julia> sequenced_reads_dna
5-element Array{Any,1}:
 ATAGCAAAAG
 ATAGCAAATAGCAATAGTTCTCACACTAGTCT
 ATAGCAAATAGCAATAGTCACAGCTAGTGCT
 ATAGCAAATAGCAATAGTTCCCACTAGTCT
 ATAGCAAATAGCAATAGTTCTCACACTAGTCT

julia> sequenced_reads_function
5-element Array{Any,1}:
 TAGCAATAG
 ATACAATA
 TCATCACTAGTCT
 TCGACACAGGTCT
 TCACACTAGTCT
```

**Comment by bicycle1885** [comment]
Actually, this is an expected behavior. Julia's [non-standard string literal](http://docs.julialang.org/en/release-0.5/manual/strings/#non-standard-string-literals) allocates an object just before runtime and the object will be persistent in memory. But calling the `DNASequence` constructor allocates an object just after it is called. Here is a reduced case:

```
julia> using Bio.Seq

julia> for _ in 1:3
           s = dna""
           push!(s, DNA_A)
           @show s
       end
s = 1nt DNA Sequence:
A
s = 2nt DNA Sequence:
AA
s = 3nt DNA Sequence:
AAA

julia> for _ in 1:3
           s = DNASequence()
           push!(s, DNA_A)
           @show s
       end
s = 1nt DNA Sequence:
A
s = 1nt DNA Sequence:
A
s = 1nt DNA Sequence:
A

```

And another case:
```julia
julia> using Bio.Seq

julia> function foo()
           s = dna""
           push!(s, DNA_A)
       end
foo (generic function with 1 method)

julia> foo()
1nt DNA Sequence:
A

julia> foo()
2nt DNA Sequence:
AA

julia> foo()
3nt DNA Sequence:
AAA

```

I also think this is confusing and should be avoided. I think we can do that by setting a shared flag when creating a sequence object with `dna""`. I will try it later.

**Comment by Ward9250** [comment]
So to make sure I understand, as I had no idea dna"" would do this - the dna_str macro just calls a BioSequence constructor does it not? So is it case that, anything created using a X_str macro then is allocated before runtime, instead of during runtime.

**Comment by bicycle1885** [comment]
You're right. Calling `copy` for each `foo` call can avoid the unintended updating.

```julia
julia> using Bio.Seq

julia> function foo()
           s = copy(dna"")
           push!(s, DNA_A)
       end
foo (generic function with 1 method)

julia> foo()
1nt DNA Sequence:
A

julia> foo()
1nt DNA Sequence:
A

julia> foo()
1nt DNA Sequence:
A

```

I thought setting the `shared` flag inside `dna_str` would solve the problem but it is wrong. `orphan!` just reallocates the `data` field of `BioSequence` and does not create a new copy of `BioSequence` itself.

A solution is deferring object creation at runtime and this can be achieved easily by defining `dna_str` as:
```julia
macro dna_str(seq, flags...)
    quote
        DNASequence($(remove_newlines(seq)))
    end
end
```
instead of:
```julia
macro dna_str(seq, flags...)
    return DNASequence(remove_newlines(seq))
end
```

This will create `DNASequence` every time when `dna"..."` is used inside a program.

**Comment by Ward9250** [comment]
I personally think after my little informal survey that the deferring behaviour you describe using:

```julia
macro dna_str(seq, flags...)
    quote
        DNASequence($(remove_newlines(seq)))
    end
end
```

would be the way to go as dna_str then behaves as everyone expected it to.

**Comment by cjprybol** [comment]
Thanks for the link and explaining the behavior Kenta! I agree with Ben that the behavior of `dna""` is confusing, but I think it's only confusing because it appeared as though `dna""` and `DNASequence` could be used interchangeably from the examples I found in the documentation.

My preference is towards making this proposed change whereby the `dna""` constructor would generate a new object by default, rather than making no change at all, just because I think it would be easy for others to make the same mistake that I did. However, with Ben's suggestion that it might be nice to have both functionalities available, one allocated before runtime and one that is deferred, maybe we could make a (rough) analogy based on the guidelines from the base language.

quoting the link that Kenta shared:

> The convention is that non-standard literals with uppercase prefixes produce actual string objects, while those with lowercase prefixes produce non-string objects like byte arrays or compiled regular expressions.

Given that convention, where lowercase non-standard string literals are compiled/byte code, I'll pitch the idea that `dna""` be left as is, meaning that all current code that works will still work, but instead add an uppercase `DNA""` constructor that would use this new proposed behavior where it always generates a new `DNASequence` object. The rough logic there is that Julia users who know about this non-standard string behavior would expect that `dna""` behaves in the efficient, yet restrictive and mistake prone manner that it is always compiled before runtime. As the base language convention is that the upper case non-standard string literals generate strings (more expensive than the byte code), assign the more expensive behavior that generates a new DNASequence object by default to an uppercase `DNA""` version of the constructor.

in summary:
- `dna""` unchanged
- add `DNA""` and make it always generate a new object, using the functions demonstrated above
- replace examples in the documentation that use `dna""` with the uppercase `DNA""`, and move the `dna""` construct into a section of the documentation targetted for advanced users and those interested in performance. And also add a warning and example of this behavior for those who aren't aware that this is intended behavior (like me)

thoughts?

**Comment by cjprybol** [comment]
I tried this to see if the performance differences were even worth caring about. They might be?

```
julia> Pkg.rm("Bio")
INFO: Removing Bio (unregistered)

julia> Pkg.clone("https://github.com/cjprybol/Bio.jl.git")
INFO: Cloning Bio from https://github.com/cjprybol/Bio.jl.git
INFO: Computing changes...

julia> using Bio.Seq
INFO: Recompiling stale cache file /Users/Cameron/.julia/lib/v0.5/Bio.ji for module Bio.

julia> using BenchmarkTools

julia> function dna_lower()
               s1 = dna""
               push!(s1, DNA_A)
       end
dna_lower (generic function with 1 method)

julia> function DNA_upper()
               s2 = DNA""
               push!(s2, DNA_A)
       end
DNA_upper (generic function with 1 method)

julia> dna_lower()
1nt DNA Sequence:
A

julia> dna_lower()
2nt DNA Sequence:
AA

julia> dna_lower()
3nt DNA Sequence:
AAA

julia> DNA_upper()
1nt DNA Sequence:
A

julia> DNA_upper()
1nt DNA Sequence:
A

julia> DNA_upper()
1nt DNA Sequence:
A

julia> @benchmark dna_lower()
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     995
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     29.00 ns (0.00% GC)
  median time:      30.00 ns (0.00% GC)
  mean time:        33.64 ns (0.00% GC)
  maximum time:     2.41 μs (0.00% GC)

julia> @benchmark DNA_upper()
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     923
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  176.00 bytes
  allocs estimate:  3
  minimum time:     110.00 ns (0.00% GC)
  median time:      115.00 ns (0.00% GC)
  mean time:        137.01 ns (11.39% GC)
  maximum time:     3.96 μs (94.52% GC)
```

I'd be happy to open a pull request if either or both of you are interested in this approach. Changes are currently in https://github.com/cjprybol/Bio.jl

**Comment by bicycle1885** [comment]
Thank you for proposing a solution. However, I've never seen that non-standard literal convention in the wild, so I'm skeptical about it. An example as far as I know is `MIME` prefix which creates a `MIME` type, not a string. I think the docs are outdated and nobody follows that convention. I'll file an issue at JuliaLang later.

An idea I come up with is using a flag of the non-standard literal. For example, `dna"ACGT"s` is **s**tatic and creates a DNA sequence only once, and `dna"ACGT"d` is **d**ynamic and creates a sequence every time. The default is static as the current code. Here is an example implementation.
```julia
macro dna_str(seq)
    # static
    return DNASequence(remove_newlines(seq))
end

macro dna_str(seq, flag)
    if "s" == flag
        # static
        return DNASequence(remove_newlines(seq))
    elseif "d" == flag
        # dynamic
        return quote
            DNASequence($(remove_newlines(seq)))
        end
    end
    error("invalid DNA flag: '$(flag)'")
end
```

**Comment by bicycle1885** [comment]
xref: https://github.com/JuliaLang/julia/issues/18898

**Comment by cjprybol** [comment]
That looks great. The flag is intuitive and makes more sense knowing that section of the language documentation is out of date.

**Comment by bicycle1885** [comment]
This is solved in #349.
